source/blog/2015-05-08-rspec-compsable-matchers.md

```diff
@@ -0,0 +1,216 @@
+**target** object is created by "expect" and basically wraps the passed object
+to it in a class called `ExpectationTarget`. The primary methods on the
+`ExpectationTarget` that users see are `to` and `not_to`/`to_not`. Let's focus
+on `to`.
+
+The `to` method is basically responsible for invoking the `matches?` method on the
+passed matcher. It's possible to see this happening quite easily in an IRB session.
+
+```
+>> require "rspec/expectations"
+=> true
+>> o = Object.new
+=> #<Object:0x007fa20d221490>
+>> o.extend(RSpec::Matchers)
+=> #<Object:0x007fa20d221490>
+>> o.expect(3).to Object.new
+NoMethodError: undefined method `matches?' for #<Object:0x007fa20d1f4e68>
+    from /Users/sam/.gem/ruby/2.1.5/gems/rspec-expectations-3.2.1/lib/rspec/expectations/handler.rb:50:in `block in handle_matcher'
+    from /Users/sam/.gem/ruby/2.1.5/gems/rspec-expectations-3.2.1/lib/rspec/expectations/handler.rb:27:in `with_matcher'
+    from /Users/sam/.gem/ruby/2.1.5/gems/rspec-expectations-3.2.1/lib/rspec/expectations/handler.rb:48:in `handle_matcher'
+    from /Users/sam/.gem/ruby/2.1.5/gems/rspec-expectations-3.2.1/lib/rspec/expectations/expectation_target.rb:54:in `to'
+    from (irb):4
+    from /Users/sam/.rubies/ruby-2.1.5/bin/irb:11:in `<main>'
+>>
+```
+
+In this session, we extend the RSpec DSL on to a blank object and then call
+`o.expect(3).to Object.new`. This blows up because our new object does not implement
+the `matches?` method.
+
+In RSpec expectations, most matchers inherit from a class called `BaseMatcher`
+which causes `matches?` to delegate to a method called `match`. Some matchers have
+a very simple definition of `match`. For example, here's the definition of `match`
+for the `eq` matcher:
+
+```ruby
+def match(expected, actual)
+  actual == expected
+end
+```
+
+Some matchers, however, have a more complicated match protocol. This more complicated
+protocol is called the **composable** matcher protocol, which is what we'll discuss next.
+
+## Composable RSpec matchers
+
+A composable RSpec matcher is one which expects other RSpec matchers for some/all of it's
+arguments. The `match` matcher is perhaps an obvious place to start. In RSpec 2.xx the match
+matcher was really only used match regexes, and its match definition looked thus:
+
+```ruby
+def match(expected, actual)
+  actual.match expected
+end
+```
+
+Now obviously, one could extend the capabilites by providing other objects that provided a
+definition of match, but that's besides the point.
+
+The new definition of the match matcher's match method is thus:
+
+```ruby
+def match(expected, actual)
+  return true if values_match?(expected, actual)
+  return false unless can_safely_call_match?(expected, actual)
+  actual.match(expected)
+end
+```
+
+The key important difference is that call to `values_match?` which is the core implementation
+of the new matching protocol in RSpec. The list of matchers which now reference the values_match?
```

Comment:
I don't consider `values_match?` to be part of any kind of new protocol.  `===` is the only additional protocol at work for composablity.  `values_match?` is just a helper method that _leverages_ the `===` protocol but isn't itself a protocol.  There's only one implementation of `values_match?` and it does something concrete...it's not an abstract interface.